admin/app/com/lucidchart/piezo/admin/utils/CronHelper.scala

```diff
@@ -0,0 +1,185 @@
+    val initialFireTime = Option(dummyTrigger.getFireTimeAfter(new Date()))
+
+    // get the interval
+    initialFireTime.fold(Long.MaxValue) { initialFireTime =>
+      logger.debug(s"sample cron expression: $simplifiedComplexCron")
+      val sampledMaxIntervalInSeconds = getSampledMaxInterval(initialFireTime, NUM_COMPLEX_SAMPLES, dummyTrigger)
+      val innerMaxIntervalSeconds = getCronContainers(secondsAndMinutes, timeZone).map(_.wrapIntervalInSeconds).sum
+      // subtract an hour to avoid double counting innerMaxIntervalSeconds
+      sampledMaxIntervalInSeconds - HourUnitType.secondsPerUnit + innerMaxIntervalSeconds
+    }
+  }
+
+  /**
+   * Estimates the max interval for a given cron expression. The more samples, the better the estimate.
+   */
+  private def getSampledMaxInterval(prev: Date, numSamples: Long, trigger: CronTrigger, maxInterval: Long = 0): Long = {
+    Option(trigger.getFireTimeAfter(prev)) match {
+      case Some(next) if numSamples > 0 =>
+        val intervalInSeconds = next.toInstant.getEpochSecond - prev.toInstant.getEpochSecond
+        if (intervalInSeconds > maxInterval) {
+          val sampleId = NUM_COMPLEX_SAMPLES - numSamples
+          logger.debug(s"Seconds:$intervalInSeconds Sample:$sampleId Interval:$prev -> $next")
+        }
+        getSampledMaxInterval(next, numSamples - 1, trigger, Math.max(intervalInSeconds, maxInterval))
+      case _ => maxInterval
+    }
+  }
+
+  private def getCronContainers(parts: Array[String], timeZone: Option[TimeZone]): List[CronContainer] = {
+    parts
+      .zip(List(SecondUnitType, MinuteUnitType, HourUnitType))
+      .map { case (str, cronType) => CronContainer(str, cronType, timeZone) }
+      .toList
+  }
+
+}
+
+sealed abstract class UnitType(val secondsPerUnit: Long, val numUnitsInContainer: Long)
+case object SecondUnitType extends UnitType(1, 60)
+case object MinuteUnitType extends UnitType(60, 60)
+case object HourUnitType extends UnitType(3600, 24)
+
+/**
+ * A cron container is composed of "units". The "unitType" determines how many units are in the container, and how many
+ * seconds are in each unit. "Marked units" describe when the trigger is set to fire. "Intervals" describe the number of
+ * units between (but not including) marked units.
+ */
+case class CronContainer(str: String, unitType: UnitType, markedUnits: List[Long], timeZone: TimeZone) {
+  lazy val areAllUnitsMarked: Boolean = markedUnits.size == unitType.numUnitsInContainer
+
+  // the "wrap interval" is the interval that wraps around both ends of the container
+  private lazy val wrapInterval = (unitType.numUnitsInContainer - markedUnits.last) + markedUnits.head
+  lazy val wrapIntervalInSeconds: Long = unitsToSeconds(wrapInterval)
+
+  // the max interval is the longest interval between any two marked units in the container, including the wrap interval
+  lazy val maxIntervalInSeconds: Long = {
```

Comment:
This currently iterates over every value. That is sub-optimal in many common cases. In particular if the expression is "*", then we can just use the duration of a single unit. If it is a range, then we just need the time between the two ends of the range. If it is an interval, then it is just the interval, possibly accounting for start time. For more complex cases, which you don't currently handle at all, it is still possible to just deal with the gaps, instead of each possible time it triggers.

Reply:
That's fair. I can make it more efficient, although we're never iterating over more than 60 items.

Reply:
> For more complex cases, which you don't currently handle at all, it is still possible to just deal with the gaps, instead of each possible time it triggers.
What do you mean by this?  Wouldn't I have to iterate over all of the values to determine the gaps? What complex cases am I not handling?

Reply:
I mention it in another comment, but things like `0-10,30-40`  or `10-40/5`